### PR TITLE
repart: Add --copy-from option

### PR DESCRIPTION
**Pull request by daandemeyer**
--copy-from synthesizes partition definitions from the given image
which are then applied to the repart algorithm. In its most basic
form, this allows copying an image to another device but it can
also be combined with --definitions to copy + add partitions in the
same call to repart.

### CODE REVIEW

File: src/partition/repart.c [hunk 1]
```diff
@@ -1801,9 +1805,230 @@ static int find_verity_sibling(Context *context, Partition *p, VerityMode mode,
...
+                uint64_t grainsz,
+                uint64_t *ret) {
+
+        size_t n_partitions;
+        uint64_t offset, next = UINT64_MAX;
+
+        assert(c);
+        assert(t);
+        assert(p);
+
```
Comment (bluca) [inline src/partition/repart.c:1845]: assert(ret)

File: src/partition/repart.c [hunk 2]
```diff
@@ -1801,9 +1805,230 @@ static int find_verity_sibling(Context *context, Partition *p, VerityMode mode,
...
+                start *= secsz;
+
+                if (partition_type_exclude(type))
+                        continue;
+
+                np = partition_new();
+                if (!np)
+                        return log_oom();
+
+                np->type = type;
```
Comment (bluca) [inline src/partition/repart.c:1991]: structured initialization maybe?

**Review by bluca — APPROVED** [review]

File: src/partition/repart.c (hunk 2 above)
Comment (daandemeyer) [reply src/partition/repart.c:1991]: That would override all the default values set in partition_new()

**Review by yuwata — APPROVED** [review]

File: src/partition/repart.c [hunk 3]
```diff
@@ -423,24 +423,20 @@ static void partition_foreignize(Partition *p) {
         p->verity = VERITY_OFF;
 }
 
-static bool partition_exclude(const Partition *p) {
-        assert(p);
-
+static bool partition_type_exclude(GptPartitionType type) {
```
Comment (poettering) [inline src/partition/repart.c:430]: hmm, copying the whole structure all the time sounds busywork. either just pass the type uuid here, or pass a pointer to the struct?

File: src/partition/repart.c [hunk 4]
```diff
@@ -423,24 +423,20 @@ static void partition_foreignize(Partition *p) {
...
+                if (sd_id128_equal(type.uuid, arg_filter_partitions[i].uuid))
                         return arg_filter_partitions_type == FILTER_PARTITIONS_EXCLUDE;
 
         return arg_filter_partitions_type == FILTER_PARTITIONS_INCLUDE;
 }
 
-static bool partition_defer(const Partition *p) {
-        assert(p);
-
+static bool partition_type_defer(GptPartitionType type) {
```
Comment (poettering) [inline src/partition/repart.c:441]: similar

File: man/systemd-repart.xml [hunk 5]
```diff
@@ -440,6 +440,14 @@
         due to missing permissions.</para></listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><option>--copy-from=</option><arg>IMAGE</arg></term>
+
+        <listitem><para>Instructs <command>systemd-repart</command> to copy the partitions from the given
+        image. The partitions from the given image are synthesized into partition definitions that are
+        parsed before the partition definition files.</para></listitem>
```
Comment (poettering) [inline man/systemd-repart.xml:448]: i cannot parse this.

"partitions … are synthesized into … definitions that are parsed"?

what's that supposed to mean?

File: src/partition/repart.c [hunk 6]
```diff
@@ -1801,9 +1805,231 @@ static int find_verity_sibling(Context *context, Partition *p, VerityMode mode,
...
+                        return log_error_errno(SYNTHETIC_ERRNO(EIO), "Partition end beyond disk end.");
+        }
+
+        assert(next >= offset);
+        offset = round_up_size(offset, grainsz);
+        next = round_down_size(next, grainsz);
+
+        *ret = LESS_BY(next, offset); /* Saturated subtraction, rounding might have fucked things up */
+        return 0;
+}
```
Comment (poettering) [inline src/partition/repart.c:1897]: these two functions are just moved up without changes? please do so in separate commits, otherwise so hard to follow what is going on

File: man/systemd-repart.xml (hunk 5 above)
Comment (poettering) [reply man/systemd-repart.xml:448]: so let me grok this: this basically copies *all* partitions from the specified source 1:1, i.e. the contents and metadata will be identical, but the offset and partno will likely be different? please say so explicitly. I had a hard time figuring out what this does.

**Comment by poettering** [comment]
So this does sound useful, but can you elaborate on your usecase?

This allows a single source currently, shouldn't it allow multiple such files to be specified?

**Comment by DaanDeMeyer** [comment]
> So this does sound useful, but can you elaborate on your usecase?

This is more or less how I envision installing an image would work. First you use --copy-from to deploy the mkosi image to your USB stick. Then you boot the USB stick. And then you use --copy-from again to deploy the installer image to the actual disk. The second --copy-from could either be with the root partition definition from the image or exclude it to have the root partition generated on first boot.

It sounds way cooler to have the installer add the root partition and to then soft-reboot into the newly installed system without rebooting though.

**Comment by poettering** [comment]
> This is more or less how I envision installing an image would work. First you use --copy-from to deploy the mkosi image to your USB stick. 

This would not really give us much above a `dd`, would it?

> Then you boot the USB stick. And then you use --copy-from again to deploy the installer image to the actual disk. The second --copy-from could either be with the root partition definition from the image or exclude it to have the root partition generated on first boot.
> 
> It sounds way cooler to have the installer add the root partition and to then soft-reboot into the newly installed system without rebooting though.

Uh, this is an interesting idea. Not sure if I love it or hate it.

**Comment by poettering** [comment]
> > 
> 
> It would make sure the partition table is resized to the full size of the USB. Linux complains about that IIRC if it isn't.

yeah, but that message can be ignored.

but maybe it's a good think, because it means we'll adjust the GPT table to the local sector size of the disk
